pattern/src/lib.rs

```diff
@@ -1,9 +1,107 @@
+//! This crate implements a macro-based DSL for matching netlists.
+//!
+//! The DSL features programmable patterns, which enables many features that
+//! wouldn't otherwise be possible. For example:
+//! - [`PZExt`] can match zero extensions, even though they don't correspond
+//!   to a cell at all, and are achieved by appropriately mixing [`Net`]s
+//!   within a [`Value`]
+//! - patterns which correspond to bitwise cells, such as [`PAnd`], can match
+//!   a [`Value`] even if it is made up from [`Net`]s coming from many distinct
+//!   [`And`] cells.
+//! - specialized patterns can be easily implemented, such as [`PPow2@a`],
+//!   which matches a constant power of 2 and captures the exponent into `a`.
+//!
+//! [`And`]: prjunnamed_netlist::Cell::And
+//! [`Net`]: prjunnamed_netlist::Net
+//! [`Value`]: prjunnamed_netlist::Value
+//! [`PAnd`]: patterns::PAnd
+//! [`PPow2@a`]: patterns::PPow2
+//! [`PZExt`]: patterns::PZExt
+//!
+//! By convention, the names of patterns are prefixed with `P`.
+//!
+//! The general netlist match syntax looks as follows:
+//!
+//! ```no_run
+//! # use prjunnamed_netlist::{Design, Net, Value};
+//! # use prjunnamed_pattern::netlist_matches;
+//! # use prjunnamed_pattern::patterns::*;
+//! let matcher = netlist_matches! {
+//!     [PAdc@y [PAny@a] [PZero] [PInput@c ("cin")]] => (y, a, c);
+//! };
+//! # let design: Design = todo!();
+//! # let target: Value = todo!();
+//! # let result: Option<(Value, Value, Net)> = matcher(&design, &target);
+//! ```
+//!
+//! A pattern call is delimited with square brackets and consists of the
+//! following parts:
+//! - the name of the pattern being called, e.g. `PAdc`
+//! - a result capture, e.g. `@y`. This is optional, and if it is not provided,
+//!   the value captured by the pattern is discarded.
+//!
+//!   Each pattern captures exactly one value. Often, it is the [`Net`] or
+//!   [`Value`] it is being matched against. In rare cases, the captured value
+//!   is of the unit type `()`, in which case it is usually discarded by the
+//!   user.
```

Comment:
Although we don't currently do this anywhere, a pattern can of course capture a tuple.